### PR TITLE
feat(cart): Shipping method adjustments

### PR DESCRIPTION
**Pull request by olivermrbl**
awaiting #6101 

-------------------

### Add shipping method adjustments
```ts
addShippingMethodAdjustments(data: CreateShippingMethodAdjustmentDTO[])
addShippingMethodAdjustments(data: CreateShippingMethodAdjustmentDTO)
addShippingMethodAdjustments(cartId: string, data: CreateShippingMethodAdjustmentDTO[], sharedContext?: Context)
```
### Set shipping method adjustments
```ts
setShippingMethodAdjustments(cartId: string, data: (CreateShippingMethodAdjustmentDTO  | UpdateShippingMethodAdjustmentDTO)[], sharedContext?: Context)
```
### Remove shipping method adjustments
```ts
removeShippingMethodAdjustments(adjustmentIds: string[], sharedContext?: Context)
removeShippingMethodAdjustments(adjustmentIds: string, sharedContext?: Context)
removeShippingMethodAdjustments(selector: Partial<ShippingMethodAdjustmentLineDTO>,sharedContext?: Context)
```

### CODE REVIEW

File: packages/cart/integration-tests/__tests__/services/cart-module/index.spec.ts [hunk 1]
```diff
@@ -697,6 +697,7 @@ describe("Cart Module Service", () => {
 
       const [method] = await service.addShippingMethods(createdCart.id, [
         {
+          cart_id: createdCart.id,
```
Comment (riqwan) [inline packages/cart/integration-tests/__tests__/services/cart-module/index.spec.ts]: q: do we need to have `createdCart.id` in 2 places?

File: packages/cart/src/models/shipping-method-adjustment.ts [hunk 2]
```diff
@@ -1,15 +1,25 @@
...
-    fieldName: "shipping_method_id",
+    onDelete: "cascade",
+    nullable: true,
+    index: "IDX_adjustment_shipping_method_id",
   })
-  shipping_method: ShippingMethod
+  shipping_method: ShippingMethod | null
+
+  @Property({ columnType: "text" })
+  shipping_method_id: string
```
Comment (riqwan) [inline packages/cart/src/models/shipping-method-adjustment.ts:22]:
```suggestion
  @Property({ columnType: "text", nullable: true })
  shipping_method_id: string | null
```

should we do this here as well + Optional fields?

File: packages/cart/src/models/shipping-method-adjustment.ts [hunk 3]
```diff
@@ -1,15 +1,25 @@
...
+  BeforeCreate,
+  Entity,
+  ManyToOne,
+  OnInit,
+  Property,
+} from "@mikro-orm/core"
 import AdjustmentLine from "./adjustment-line"
 import ShippingMethod from "./shipping-method"
 
 @Entity({ tableName: "cart_shipping_method_adjustment_line" })
```
Comment (riqwan) [inline packages/cart/src/models/shipping-method-adjustment.ts]:
```suggestion
@Entity({ tableName: "cart_shipping_method_adjustment" })
```

File: packages/cart/src/services/cart-module.ts [hunk 4]
```diff
@@ -681,4 +689,200 @@ export default class CartModuleService implements ICartModuleService {
...
+    const cart = await this.retrieve(
+      cartId,
+      { select: ["id"], relations: ["shipping_methods.adjustments"] },
+      sharedContext
+    )
+
+    const methodIds = cart.shipping_methods?.map((method) => method.id)
+
+    const existingAdjustments = await this.listShippingMethodAdjustments(
+      { shipping_method_id: methodIds },
```
Comment (riqwan) [inline packages/cart/src/services/cart-module.ts]: `methodIds` here could be undefined, I wonder if this method returns all adjustments here in that case?

File: packages/cart/src/services/cart-module.ts [hunk 5]
```diff
@@ -681,4 +689,200 @@ export default class CartModuleService implements ICartModuleService {
...
+        if (!adjustmentsSet.has(adj.id)) {
+          toDelete.push(adj)
+        }
+      }
+    )
+
+    await this.shippingMethodAdjustmentService_.delete(
+      toDelete.map((adj) => adj!.id),
+      sharedContext
+    )
```
Comment (riqwan) [inline packages/cart/src/services/cart-module.ts]:
```suggestion
    if (toDelete.length) {
      await this.shippingMethodAdjustmentService_.delete(
        toDelete.map((adj) => adj!.id),
        sharedContext
      )
    }
```

File: packages/cart/src/services/cart-module.ts [hunk 6]
```diff
@@ -681,4 +689,200 @@ export default class CartModuleService implements ICartModuleService {
...
+      populate: true,
+    })
+  }
+
+  async addShippingMethodAdjustments(
+    adjustments: CartTypes.CreateShippingMethodAdjustmentDTO[]
+  ): Promise<CartTypes.ShippingMethodAdjustmentDTO[]>
+  async addShippingMethodAdjustments(
+    adjustment: CartTypes.CreateShippingMethodAdjustmentDTO
+  ): Promise<CartTypes.ShippingMethodAdjustmentDTO[]>
```
Comment (riqwan) [inline packages/cart/src/services/cart-module.ts]:
```suggestion
  ): Promise<CartTypes.ShippingMethodAdjustmentDTO>
```

Discard away if this is intentional, but it does look lil weird that it returns an array when the input is not

File: packages/types/src/cart/service.ts [hunk 7]
```diff
@@ -136,4 +139,44 @@ export interface ICartModuleService extends IModuleService {
...
+    config?: FindConfig<ShippingMethodAdjustmentDTO>,
+    sharedContext?: Context
+  ): Promise<ShippingMethodAdjustmentDTO[]>
+
+  addShippingMethodAdjustments(
+    data: CreateShippingMethodAdjustmentDTO[]
+  ): Promise<ShippingMethodAdjustmentDTO[]>
+  addShippingMethodAdjustments(
+    data: CreateShippingMethodAdjustmentDTO
+  ): Promise<ShippingMethodAdjustmentDTO[]>
```
Comment (riqwan) [inline packages/types/src/cart/service.ts]:
```suggestion
  ): Promise<ShippingMethodAdjustmentDTO>
```

if the other comment above tracks

File: packages/types/src/cart/service.ts [hunk 8]
```diff
@@ -136,4 +139,44 @@ export interface ICartModuleService extends IModuleService {
...
+    )[],
+    sharedContext?: Context
+  ): Promise<ShippingMethodAdjustmentDTO[]>
+
+  removeShippingMethodAdjustments(
+    adjustmentIds: string[],
+    sharedContext?: Context
+  ): Promise<void>
+  removeShippingMethodAdjustments(
+    adjustmentIds: string,
```
Comment (riqwan) [inline packages/types/src/cart/service.ts]:
```suggestion
    adjustmentId: string,
```

**Review by riqwan — APPROVED** [review]
added some comments, but lgtm!

File: packages/cart/integration-tests/__tests__/services/cart-module/index.spec.ts (hunk 1 above)
Comment (olivermrbl) [reply packages/cart/integration-tests/__tests__/services/cart-module/index.spec.ts]: Nope you're right, will create a separate type 👍

**Pull request merged**